bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerEntries.java

```diff
@@ -0,0 +1,102 @@
+/*
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.client;
+
+import io.netty.util.Recycler;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.List;
+
+public class LedgerEntries implements org.apache.bookkeeper.client.api.LedgerEntries {
+    private List<LedgerEntry> entries;
+    private final Recycler.Handle<LedgerEntries> recyclerHandle;
+
+    private LedgerEntries(Recycler.Handle<LedgerEntries> recyclerHandle) {
+        this.recyclerHandle = recyclerHandle;
+    }
+
+    private static final Recycler<LedgerEntries> RECYCLER = new Recycler<LedgerEntries>() {
+        @Override
+        protected LedgerEntries newObject(Recycler.Handle<LedgerEntries> handle) {
+            return new LedgerEntries(handle);
+        }
+    };
+
+    private void recycle() {
+        releaseByteBuf();
+        recyclerHandle.recycle(this);
+    }
+
+    private void releaseByteBuf() {
+        if (entries != null) {
+            for (LedgerEntry entry : entries) {
+                entry.getEntryBuffer().release(1);
+            }
+        }
+    }
+
+    /**
+     * Create ledger entries.
+     *
+     * @param entries the entries with ordering
+     * @return the LedgerEntries
+     */
+    public static LedgerEntries create(Enumeration<LedgerEntry> entries) {
+        LedgerEntries ledgerEntries = RECYCLER.get();
+        ledgerEntries.entries = Collections.list(entries);
+        return ledgerEntries;
+    }
+
+    @Override
+    public LedgerEntry getEntry(long entryId) {
+        long firstId = entries.get(0).getEntryId();
+        long lastId = entries.get(entries.size() - 1).getEntryId();
+        if (entryId < firstId || entryId > lastId) {
+            return null;
+        }
+        return entries.get((int)(entryId - firstId));
+    }
+
+    @Override
+    public Iterator<org.apache.bookkeeper.client.api.LedgerEntry> iterator() {
```

Comment:
I think we have to rethink on the `LedgerEntry` interface as well. The `LedgerEntry` interface is completely a mess. `getEntry()` and `getEntryBuffer()` have completely different behaviors. `getEntry()` releases bytebuf automatically, while `getEntryBuffer()` returns the bytebuf (if you don't call getEntry, you are responsible for releasing the entry buffer.

I would suggest:

1) removing `getEntry()` from LedgerEntry, or make it reenterable. 
2) make LedgerEntry implement AutoCloseable. LedgerEntry should be responsible for releasing the bytebuf it is holding.
3) LedgerEntry should implement duplicate method. `iterator` here should have a separate copy of LedgerEntry and make a slice of the underlying bytebuf.

@jiazhai @eolivelli

Reply:
Good ideas

Reply:
Good ideas

Reply:
+1, opened #731 for this idea.